File: Compiler/src/parsing/lexing/Token.cpp
#include "parsing/lexing/Token.hpp"
#include <cassert>

//TODO: Huge performance boost potential
const char* TOKEN_TEXT[] = {
	"pub", "let", "def", "typedef", "namedef", "with", "as", "mut", "uncrt", "certain", "move",

	"class", "trait", "enum", "prot",
	"ctor", "dtor",  "this", //own
	"inline", "virt", "override",

	"if", "else", "elselse", "for", "while", "do", "match",
	"continue", "break", "retry", "return",

	"char",
	"i8", "u8", "i16", "u16", "i32", "u32",
	"i64", "u64", "usize", "isize", "bool", "f32", "f64",

	//	"shared", "unique", "new", "delete",

	"size_of", "type_of", "length_of"

	"true", "false", "null"
};

//This could be packed together so nicely
const char* ONE_CHAR_TOKEN_TEXTS[] = {
	"=", ":", ";",
	"(", ",", ")",
	"{", "}", ".",
	"@", "[", "]",
	"$",

	"+", "-", "*", "/", "%",
	"&", "|", "^", "!",
	"~", "<", ">", "?"
};

const char* COMPOSITE_TOKENS[] = {
	":=", "::",
	"<<", ">>", ">>>", "&&", "||",
	"==", "!=", ">=", "<=", "++", "--",

	"&mut", "&move", "&unique", "&shared"
};

const char* TEXT_TOKENS[] = {
	"Identifier", "String literal", "Char literal",
	"Integer literal", "Long literal", "Float literal", "Double literal"
};

struct TokenMerge {
	TokenType first;
	TokenType second;
	TokenType result;
	TokenMerge(TokenType p_first, TokenType p_second, TokenType p_result) : first(p_first), second(p_second), result(p_result) {}
};

TokenMerge TOKEN_MERGES[] = {
	TokenMerge(TYPE_SEPARATOR,ASSIGN,DECLARE), TokenMerge(TYPE_SEPARATOR, TYPE_SEPARATOR, MODULE_ACCESS),
	TokenMerge(LOWER, LOWER, LSL), TokenMerge(GREATER, GREATER, ASR), TokenMerge(ASR, GREATER, LSL),
	TokenMerge(REF, REF, LOGICAL_AND), TokenMerge(BITWISE_OR, BITWISE_OR, LOGICAL_OR),
	TokenMerge(ASSIGN, ASSIGN, EQUALS), TokenMerge(NOT, ASSIGN, NOT_EQUALS),
	TokenMerge(GREATER, ASSIGN, GREATER_OR_EQUAL), TokenMerge(LOWER, ASSIGN, LOWER_OR_EQUAL),
	TokenMerge(PLUS, PLUS, PLUS_PLUS), TokenMerge(MINUS, MINUS, MINUS_MINUS),

	TokenMerge(REF, MUT, MUT_REF)
	//, TokenMerge(REF, MOVE, MOVE_REF),
	//TokenMerge(REF, UNIQUE, UNIQUE_PTR), TokenMerge(REF, SHARED, SHARED_PTR)
};

const char* getTokenTypeText(const TokenType& type) {
	if(type < (sizeof(TOKEN_TEXT)/sizeof(*TOKEN_TEXT)))
		return TOKEN_TEXT[type];
	else if((unsigned)type-FIRST_ONE_CHAR_TOKEN < (sizeof(ONE_CHAR_TOKEN_TEXTS)/sizeof(*ONE_CHAR_TOKEN_TEXTS)))
		return ONE_CHAR_TOKEN_TEXTS[type-FIRST_ONE_CHAR_TOKEN];
	else if((unsigned)type-FIRST_COMPOSITE_TOKEN < (sizeof(COMPOSITE_TOKENS)/sizeof(*COMPOSITE_TOKENS)))
		return COMPOSITE_TOKENS[type-FIRST_COMPOSITE_TOKEN];
	else if((unsigned)type-FIRST_TEXT_TOKEN < (sizeof(TEXT_TOKENS)/sizeof(*TEXT_TOKENS)))
		return TEXT_TOKENS[type-FIRST_TEXT_TOKEN];

	switch(type) {
	case END_TOKEN:
		return "EOF Token";
	case ERROR_TOKEN:
		return "Error Token";
	default:
		return "Unknown token, somehow";
	}
}

const char* getTokenText(const Token& token) {
	if(token.type >= FIRST_TEXT_TOKEN && token.type <= LAST_TEXT_TOKEN) {
		if(token.text.size()==0)
			return "_";
		return token.text.c_str();
	}
	return getTokenTypeText(token.type);
}

Token::Token() : type(NEVER_SET_TOKEN), text(), literalKind(LiteralKind::I8), integer(0), real(0.0), line(0), col(0), endCol(0) {}

void resetTokenSetText(Token& token, const std::string& text) {
	token.text = text;
	token.real = 0.0;
	token.integer = 0;
}

void resetTokenSpecialValues(Token& token) {
	//token.text.clear(); //Optimize: Find out if this is bad for performance
	token.real = 0.0;
	token.integer = 0;
}

bool setTokenFromOwnWord(Token& token, int line, int startCol, int endCol) {
	token.line = line;
	token.col = startCol;
	token.endCol = endCol;
	//Optimize: This atrocity
	for(unsigned int tokenType = 0; tokenType < sizeof(TOKEN_TEXT)/sizeof(char*); tokenType++) {
		if(token.text==TOKEN_TEXT[tokenType]) {
			token.type = static_cast<TokenType>(tokenType);
			resetTokenSpecialValues(token);
			return true;
		}
	}

	token.type = IDENTIFIER;
	if(token.text == "_")
		token.text.clear(); //A size 0 identifier
	//The text is already set
	return true;
}

bool setTokenFromSpecialChar(Token& token, char c, int line, int col) {

	for(unsigned int i = 0; i < (sizeof(ONE_CHAR_TOKEN_TEXTS)/sizeof(char*)); i++) {
		if(ONE_CHAR_TOKEN_TEXTS[i][0]==c) {
			token.type = static_cast<TokenType>(i+FIRST_ONE_CHAR_TOKEN);
			resetTokenSpecialValues(token);
			token.line = line;
			token.col = col;
			token.endCol = col+1;
			return true;
		}
	}
	return false;
}

void setTokenFromRealNumber(Token& token, LiteralKind realType, daf_largest_float real, int line, int col, int endCol, const std::string& text) {
	token.type = REAL_LITERAL;
	token.real = real;
	token.literalKind = realType;
	token.integer = 0;
	token.line = line;
	token.col = col;
	token.endCol = endCol;
	token.text = text;
}

void setTokenFromInteger(Token& token, LiteralKind intType, daf_largest_uint integer, int line, int col, int endCol, const std::string& text) {
	token.type = INTEGER_LITERAL;
	token.integer = integer;
	token.literalKind = intType;
	token.real = 0.0;
	token.line = line;
	token.col = col;
	token.endCol = endCol;
	token.text = text;
}

void setTokenFromStringLiteral(Token& token, std::string&& text, int line, int col, int endLine, int endCol) {
	token.type = STRING_LITERAL;
	token.text = std::move(text);
	token.real = 0.0;
	token.integer = 0;
	token.line = line;
	token.col = col;
	token.endCol = endCol;
	assert(line == endLine); //TODO: Support multiline tokens
}

bool mergeTokens(Token& first, const Token& second) {
	if(first.line != second.line || first.endCol != second.col) //You can't put a comment inside a token
		return false;
	for(unsigned int i = 0; i < sizeof(TOKEN_MERGES)/sizeof(*TOKEN_MERGES); i++) {
		TokenMerge& merg = TOKEN_MERGES[i];
		if(merg.first == first.type && merg.second == second.type) {
			first.type = merg.result;
			first.endCol = second.endCol;
			return true;
		}
	}
	return false;
}

void setProperEOFToken(Token& token, int line, char col) {
	resetTokenSpecialValues(token);
	token.type = END_TOKEN;
	token.line = line;
	token.col = col;
	token.endCol = col;
}

bool isTokenPrimitive(TokenType type) {
	return type >= FIRST_PRIMITVE_TOKEN && type <= LAST_PRIMITIVE_TOKEN;
}

File: OCompiler/src/token.ml

type token =
  | Pub | Let | Def | With | As | Mut | Uncrt | Move | Copy
  | Class | Trait | Namespace | Enum | Prot
  | Ctor | Dtor | This | This_Type
  | Virt | Override

  | If | Else | For | While | Do | Match
  | Continue | Break | Retry | Return

  | Char
  | I8 | U8 | I16 | U16 | I32 | U32 | I64 | U64
  | Usize | Isize | Bool | F32 | F64

  | Sizeof | Typeof | Lengthof
  | True | False | Null

  | Assign | Type_Separator | Statement_End | Left_Paren | Comma | Right_Paren
  | Scope_Start | Scope_End | Left_Bracket | Right_Bracket | Type_Infered

  | Plus | Minus | Mult | Divide | Modulo
  | Lsl | Asr | Lsr
  | Ref | Dereference | Class_Access
  | Bitwise_Or | Bitwise_Xor | Not | Bitwise_Not
  | Logical_And | Logical_Or
  | Equals | Not_Equals | Greater_Or_Equal | Lower_Or_Equal | Lower | Greater
  | Q_mark

  | Plus_Plus | Minus_Minus

  | Identifier of string | String_Literal of string | Integer_Literal of int | Real_Literal of float
  | Error of char

let token_to_string token =
  match token with
  | Pub -> "pub" | Let -> "let" | Def -> "def" | With -> "with" | As -> "as" | Mut -> "mut" | Uncrt -> "uncrt" | Move -> "move" | Copy -> "copy"
  | Class -> "class" | Trait -> "trait" | Namespace -> "namespace" | Enum -> "enum" | Prot -> "prot"
  | Ctor -> "ctor" | Dtor -> "dtor" | This -> "this" | This_Type -> "This"
  | Virt -> "virt" | Override -> "override"

  | If -> "if" | Else -> "else" | For -> "for" | While -> "while" | Do -> "do" | Match -> "match"
  | Continue -> "continue" | Break -> "break" | Retry -> "retry" | Return -> "return"

  | Char -> "char"
  | I8 -> "i8" | U8 -> "u8" | I16 -> "i16" | U16 -> "u16" | I32 -> "i32" | U32 -> "u32" | I64 -> "i64" | U64 -> "u64"
  | Usize -> "usize" | Isize -> "isize" | Bool -> "bool" | F32 -> "f32" | F64 -> "f64"

  | Sizeof -> "sizeof" | Typeof -> "typeof" | Lengthof -> "lengthof"
  | True -> "true" | False -> "false" | Null -> "null"

  | Assign -> "=" | Type_Separator -> ":" | Statement_End -> ";" | Left_Paren -> "(" | Comma -> "," | Right_Paren -> ")"
  | Scope_Start -> "{" | Scope_End -> "}" | Left_Bracket -> "[" | Right_Bracket -> "]" | Type_Infered -> "$"

  | Plus -> "+" | Minus -> "-" | Mult -> "*" | Divide -> "/" | Modulo -> "%"
  | Lsl -> "<<" | Asr -> ">>" | Lsr -> ">>>"
  | Ref -> "&" | Class_Access -> "." | Dereference -> "@"
  | Bitwise_Or -> "|" | Bitwise_Xor -> "^" | Not -> "!" | Bitwise_Not -> "~"
  | Logical_And -> "&&" | Logical_Or -> "||"
  | Equals -> "==" | Not_Equals -> "!=" | Greater_Or_Equal -> ">=" | Lower_Or_Equal -> "<=" | Lower -> "<" | Greater -> ">"
  | Q_mark -> "?"

  | Plus_Plus -> "++" | Minus_Minus -> "--"

  | Identifier string -> String.concat string ["_";"_"]
  | String_Literal string -> String.concat string ["\""; "\""]
  | Integer_Literal int -> string_of_int int
  | Real_Literal float -> string_of_float float
  | Error char -> Printf.sprintf "ERROR_TOKEN %c" char


type loc_t = {line:int; col:int}
type span_t = {loc:loc_t; len:int;}
type token_with_span = (token * span_t)

let span (loc:loc_t) (len:int) :span_t = {loc=loc;len=len}

let span_loc span = span.loc
let span_end span = {line=span.loc.line; col=span.loc.col+span.len}
let span_combine head tail = {loc=head.loc; len=tail.loc.col+tail.len-head.loc.col}

let string_to_token (text : string) (loc : loc_t) : token_with_span =
  let tok = match text with
  | "pub" -> Pub | "let" -> Let | "def" -> Def | "with" -> With | "as" -> As | "mut" -> Mut | "uncrt" -> Uncrt | "move" -> Move | "copy" -> Copy
  | "class" -> Class | "trait" -> Trait | "namespace" -> Namespace | "enum" -> Enum | "prot" -> Prot
  | "ctor" -> Ctor | "dtor" -> Dtor | "this" -> This | "This" -> This_Type
  | "virt" -> Virt | "override" -> Override

  | "if" -> If | "else" -> Else | "for" -> For | "while" -> While | "do" -> Do | "match" -> Match
  | "continue" -> Continue | "break" -> Break | "retry" -> Retry | "return" -> Return

  | "char" -> Char
  | "i8" -> I8 | "u8" -> U8 | "i16" -> I16 | "u16" -> U16 | "i32" -> I32 | "u32" -> U32 | "i64" -> I64 | "u64" -> U64
  | "usize" -> Usize | "isize" -> Isize | "bool" -> Bool | "f32" -> F32 | "f64" -> F64

  | "sizeof" -> Sizeof | "typeof" -> Typeof | "lengthof" -> Lengthof
  | "true" -> True | "false" -> False | "null" -> Null
  | id -> Identifier id in
  (tok, {loc=loc; len=String.length text})

let char_to_token (c : char) (loc : loc_t) : token_with_span =
  let tok = match c with
  | '=' -> Assign | ':' -> Type_Separator | ';' -> Statement_End | '(' -> Left_Paren | ',' -> Comma | ')' -> Right_Paren
  | '{' -> Scope_Start | '}' -> Scope_End | '[' -> Left_Bracket | ']' -> Right_Bracket | '$' -> Type_Infered

  | '+' -> Plus | '-' -> Minus | '*' -> Mult | '/' -> Divide | '%' -> Modulo
  | '&' -> Ref | '.' -> Class_Access | '@' -> Dereference
  | '|' -> Bitwise_Or | '^' -> Bitwise_Xor | '!' -> Not | '~' -> Bitwise_Not
  | '<' -> Lower | '>' -> Greater
  | '?' -> Q_mark
  | _ -> Error c in
  (tok, {loc=loc; len=1})

type merge_t = {input:token*token; output:token}
let merges = [
  {input=(Assign,     Assign);     output=Equals};
  {input=(Not,        Assign);     output=Not_Equals};
  {input=(Lower,      Lower);      output=Lsl};
  {input=(Greater,    Greater);    output=Asr};
  {input=(Asr,        Greater);    output=Lsr};
  {input=(Ref,        Ref);        output=Logical_And};
  {input=(Bitwise_Or, Bitwise_Or); output=Logical_Or};
  {input=(Lower,      Assign);     output=Lower_Or_Equal};
  {input=(Greater,    Assign);     output=Greater_Or_Equal};
  {input=(Plus,  Plus);  output=Plus_Plus};
  {input=(Minus, Minus); output=Minus_Minus};]

let get_merge (tok1, tok1_span) (tok2, tok2_span) =
  if (span_end tok1_span) <> (span_loc tok2_span) then None else
    let rec check_list list =
      match list with
      | {input=(match1, match2); output=out} :: rest ->
        if match1 == tok1 && match2 == tok2 then Some (out, span_combine tok1_span tok2_span) else check_list rest
      | [] -> None
    in
    check_list merges

let merge_tokens_in_stream stream =
  let rec maybe_merge_with_one tok1:token_with_span =
    match Stream.peek stream with
    | Some tok2 -> (
      match get_merge tok1 tok2 with
        | Some merge ->
          Stream.junk stream; (* Eat tok2 *)
          maybe_merge_with_one merge
        | None -> tok1 (* tok2 is still in stream *)
    )
    | None -> tok1 in
  let next _ =
    match Stream.peek stream with
    | Some tok1 ->
      Stream.junk stream;
      Some (maybe_merge_with_one tok1)
    | None -> None
  in Stream.from (next)
